Danstough/ch28072/config option for inputting an array of fields (#1878)
replicatedhq/kots: design/variadic_config_alt.md

# Variadic Config Proposal

Vendors require the ability to dynamically create resources as part of install configuration.

> One common use case is installing operators, where the customers need to create dynamic resources, like instances of an application or service, that are unknown until install time.

Reviewer: ```suggestion
One common use case is installing operators, where the customers need to create dynamic resources, like instances of an application or service, the number of which is unknown until install time.
```

1. Repeatable Config Groups

The first consideration is how the revised API will look to Vendors using these features in their application.

### Example Revised Kotskind Resources 

Vendors will use the revised Config Spec to define templateGroups and repeatable Config Items. 
Values are inserted by the Kots and returned as part of the API for creating the ConfigValues spec.
Below is a representative resource.

#### Config
```yaml
apiVersion: kots.io/v1beta1 
kind: Config 
metadata: 
  creationTimestamp: null 
  name: config-sample 
spec: 
  # UNCHANGED
  groups: 
  - name: nginx_settings 
    title: Nginx Configs 
    description: Config to serve as an example for creating your own
    items: 
    - name: "nginx_port"
      type: "text"
      title: "Nginx port"
      default: "80"
      value: ""

    # NEW!
    # This is a repeatable Config Item
    - name: "static_files"
      type: file
      title: "Static Assets"
      repeatable: true      # NEW! Tells the UI/Kots to expect an array
      minimumCount: 3       # NEW! Not sure if this is needed here but including for discussion
      repeatValues:         # NEW! Returned to the API filled in from the CLI/console

>       - "encoded file value one"

Reviewer 2: I think that we will need to have some kind of "ID" as the other repeatValues in the example to have the possibility to add the filename from an end user perspective.

Reviewer 3: This has been added in with the updated design

      value: bravo
```

### Resource Templates

In addition to using the new syntax in the KOTS CRDS, they will template their resources using Golang Text Template `range` syntax like the following examples.
Comments are used to keep valid syntax for linting purposes and also provide explicit documentation of generated fields.

#### New ConfigContext Methods

| Method            | Input | Output | Purpose |
|-------------------|-------|--------|---------|
| ConfigOptionList  |       |        |         |
| ConfigOptionIndex |       |        |         |
| ConfigGroupList   |       |        |         |
|                   |       |        |         |
|                   |       |        |         |

#### Repeatable Config Item Usage

Mounting a bunch of secrets (files) to a container as config data.

```yaml
# deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-deploy
spec:
  template:
    spec:
      containers:
        - name: test
          image: httpd
          volumeMounts:
            - mountPath: /var/www/html
              name: secret-assets
              readOnly: true
      volumes:
      - name: secret-assets
        projected:
          sources:

>           # repl{{ range $index, $value := ConfigOptionList "static_files" }}

Reviewer 2: From an end user perspective we need to have a way to "select" what filename the static_file should be exposed as, for example file1: index.html, file2: favicon.ico ...
